homeassistant/components/websocket_api/commands.py

```diff
@@ -225,7 +229,35 @@ def handle_get_states(
             if entity_perm(state.entity_id, "read")
         ]
 
-    connection.send_result(msg["id"], states)
+    # JSON serialize here so we can recover if it blows up due to the
+    # state machine containing unserializable data. This command is required
+    # to succeed for the UI to show.
+    response = messages.result_message(msg["id"], states)
+    try:
+        connection.send_message(const.JSON_DUMP(response))
+        return
+    except (ValueError, TypeError):
+        connection.logger.error(
+            "Unable to serialize to JSON. Bad data found at %s",
+            format_unserializable_data(
+                find_paths_unserializable_data(response, dump=const.JSON_DUMP)
+            ),
+        )
+    del response
+
+    # If we can't serialize, we'll filter out unserializable states
+    serialized = []
+    for state in states:
+        try:
+            serialized.append(const.JSON_DUMP(state))
+        except (ValueError, TypeError):
+            # Error is already logged above
+            pass
+
+    # We now have partially serialized states. Craft some JSON.
+    response2 = const.JSON_DUMP(messages.result_message(msg["id"], ["TO_REPLACE"]))
+    response2 = response2.replace('"TO_REPLACE"', ", ".join(serialized))
```

Comment:
I really wish the python json lib would allow constructing json from already serialized strings.  It would open up the opportunity to make things so much more efficient

Reply:
I guess we could add

```python
class JSONStr(str):
    pass
```

And then teach our `JSONEncoder` that instances of `JSONStr` can be used as-is

Reply:
Yeah thats way cleaner, and some other things can be sped up as well

Reply:
Was looking at the JSONEncoder implementation and it's doesn't allow overriding custom serialization. It only allows converting unsupported objects to supported objects.

Reply:
I don't see a way of making it work either.  https://github.com/python/cpython/blob/main/Lib/json/encoder.py#L439 just feeds back into itself